Adjust neural network reductor to run without full order model

TiPlath opened the pull request:
I use pyMOR to reduce the order of discrete particle simulation. These simulation have no governing equation and are computationally very expensive. Thus I want to use the data-driven neural network approach with my simulation data. I noticed some bugs in the implementation when no ```fom``` is given. this pull request should resolve these and also (I hope) tidy up the implementation a bit.

Main changes:

- Split training and validation data into a set of parameter values and corresponding solution snapshots. ```training_snapshots``` and ```validation_snapshots``` were added as input parameter, but can still be computed if none are given and a ```fom``` is available
- Split up ```compute_training_data()``` into multiple functions: ```compute_training_snapshots()```, ```compute_reduced_basis()``` and ```compute_training_data```
- Adjusted the ```neural_networks.py``` demo to account for the changes
- Added assertions to check for errors
- Added computation of ```validation_snapshots```
- Added documentation for these changes.

I have done some rigorous testing, but might have missed something with my current (limited) knowledge of pyMOR. Feel free to point me towards adjustments and improvements, they are very welcome.

TiPlath commented:
MWE for my data: [simulationStudy.zip](https://github.com/user-attachments/files/16113269/simulationStudy.zip)

Parameters are fictional and non-physical this is at least my explanation for the high error values of the reductor.

```
@skip_if_missing('TORCH')
def test_no_fom():
    from pymor.reductors.neural_network import NeuralNetworkReductor
    from pymor.core import pickle
    from pymor.parameters.base import Parameters
    from pymor.vectorarrays.numpy import NumpyVectorSpace
    from pymor.models.neural_network import NeuralNetworkModel

    density_matrix = pickle.load(open("data/simulationStudy.pickle", "rb"))

    parameters = Parameters({"volumeRatio": 1})
    training_set = [parameters.parse(p) for p in [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0]]
    training_snapshots = density_matrix[:, :, :, -1]
    training_snapshots = NumpyVectorSpace.from_numpy(training_snapshots.reshape((training_snapshots.shape[0], -1)))

    reductor_data_driven = NeuralNetworkReductor(training_set=training_set, training_snapshots=training_snapshots,
                                                 l2_err=1e-2, ann_mse=1e-2)
    rom_data_driven = reductor_data_driven.reduce(restarts=100, log_loss_frequency=10)

    # Check that the reduced order model exists
    assert rom_data_driven

    # Check that the reduced order model is of the correct type
    assert isinstance(rom_data_driven, NeuralNetworkModel)

    # Check the dimension of the reduced order model
    assert rom_data_driven.solution_space.dim <= training_snapshots.dim
```

HenKlei commented:
Hi @TiPlath, thanks for contributing to pyMOR!

Let me start with two questions: Are the `training_snapshots` supposed to be a single `VectorArray` or a list of `VectorArray`s? Further, while adjusting the code, did you also have the instationary reductors in mind? I think with the current structure of your changes these might cause trouble.

TiPlath commented:
> Hi @TiPlath, thanks for contributing to pyMOR!
> 
> Let me start with two questions: Are the `training_snapshots` supposed to be a single `VectorArray` or a list of `VectorArray`s? Further, while adjusting the code, did you also have the instationary reductors in mind? I think with the current structure of your changes these might cause trouble.

Hi @HenKlei, 
The ```training_snapshots``` are ```NumpyVectorArrays``` with ```N``` snapshots which are represented by a vector. The assembled ```training_data``` however is a list of tuples with the ```training_set``` and ```training_snapshots``` as tuple contents.
I saw the instationary ones failing and I am currently working on fixing this. stay tuned ;)

TiPlath commented:
> > Hi @TiPlath, thanks for contributing to pyMOR!
> > Let me start with two questions: Are the `training_snapshots` supposed to be a single `VectorArray` or a list of `VectorArray`s? Further, while adjusting the code, did you also have the instationary reductors in mind? I think with the current structure of your changes these might cause trouble.
> 
> Hi @HenKlei, The `training_snapshots` are `NumpyVectorArrays` with `N` snapshots which are represented by a vector. The assembled `training_data` however is a list of tuples with the `training_set` and `training_snapshots` as tuple contents. I saw the instationary ones failing and I am currently working on fixing this. stay tuned ;)

@HenKlei, now the instationary neural networks work as well. However, it somehow got bigger than I planned it to be. I deleted some functions which were (in my eyes) code duplication and merged the instationary and stationary computation of the ```training_data``` and ```validation_data```. If I changed too much and in your eyes this does not fit the code style, we can reverse this. I also see the point of having different functions for stationary and instationary which makes it more sorted somehow.

HenKlei commented:
Thanks @TiPlath, I will take a closer look next week. Removing duplicate code is always a good idea and I also had the impression already before that there is some potential for improvement and unification of the stationary and instationary cases.

HenKlei reviewed (commented):
In general, I like the new approach a lot. I am still not really sure how to properly treat the instationary case. Maybe we can schedule a short meeting to discuss those details @TiPlath? If that is fine for you, just drop me an email.

TiPlath commented on src/pymor/reductors/neural_network.py:
```diff
@@ -245,44 +262,69 @@ def _initialize_neural_network(self, params):
             self.training_data = []
-            for mu, u in training_set_iterable:
-                sample = self._compute_sample(mu, u)
+            for i, (mu, u) in enumerate(training_data_iterable):
+                samples = self._compute_sample(mu, u.base[i*self.nt:(i+1)*self.nt])
```
simplified the function

TiPlath commented on src/pymor/reductors/neural_network.py:
```diff
@@ -767,143 +736,23 @@ def _compute_layer_sizes(self, hidden_layers):
     This is a reductor that trains an LSTM neural network that approximates
     the mapping from parameter space to output space.
     """
 
```
Not sure if these two lines here are needed. I think they are never called, could that be true? If you confirm  we can delete or at least comment them out as well.

TiPlath commented on src/pymor/reductors/neural_network.py:
```diff
@@ -767,143 +736,23 @@ def _compute_layer_sizes(self, hidden_layers):
-
-        return sample
+    _initialize_neural_network = NeuralNetworkLSTMReductor._initialize_neural_network
 
```
Not sure if these two lines above here are needed. I think they are never called, could that be true? If you confirm we can delete or at least comment them out as well.

TiPlath replied:
there should also be a *.reduce line here, not sure why it is not visible

HenKlei replied:
I need to think about how the inheritance in this case actually works.

HenKlei commented:
> @HenKlei the instationary classes are removed now, we were right that they were redundant as soon as we have the is_stationary boolean. I merged all the functionalities of the classes together into a single class.
> 
> I think I am done with all the changes now. Happy to receive some feedback as discussed.

Thanks @TiPlath, I will take a look as soon as possible!

HenKlei reviewed (changes requested):
Could you briefly explain to me why you would like to split training and validation data already when passing them to the reductors? I do not see any benefit in doing so compared to splitting the data into a training and a validation set within the reductor.

HenKlei commented:
There is now one month left in order to get this into the 2024.2 release. Can we manage to finish this until then @TiPlath? Let me know if you need any help. I can also take care of some of the implementation as soon as we fix how to proceed here.

TiPlath commented:
> There is now one month left in order to get this into the 2024.2 release. Can we manage to finish this until then @TiPlath? Let me know if you need any help. I can also take care of some of the implementation as soon as we fix how to proceed here.

Hi Hendrik, 

thanks for mentioning this, that pushes it up on my priority list ;) I can take a look during the weekend. If i do not manage to get everything done at the weekend, I am happy if you take over to run the last mile. Hope to finish it to a point where it is only cosmetics you have to deal with. Currently very busy with the thesis, deadline is 4th Dec.

Timo

TiPlath commented on src/pymor/reductors/neural_network.py:
```diff
@@ -89,9 +93,9 @@ class NeuralNetworkReductor(BasicObject):
-    def __init__(self, fom=None, training_set=None, validation_set=None, validation_ratio=0.1,
-                 basis_size=None, rtol=0., atol=0., l2_err=0., pod_params={},
-                 ann_mse='like_basis', scale_inputs=True, scale_outputs=False):
+    def __init__(self, fom=None, reduced_basis=None, training_set=None, validation_set=None, training_snapshots=None,
```
> The documentation of `reduced_basis` is missing if I am not mistaken.
> 
> Please also add this addition in the first comment of this pull request such that we can find it easier later on.

what do you mean by "add this addition to the first comment of this PR?" Not sure what to do

TiPlath replied:
got it now by reading further

TiPlath commented on src/pymor/reductors/neural_network.py:
```diff
@@ -441,8 +521,9 @@ class NeuralNetworkStatefreeOutputReductor(NeuralNetworkReductor):
 
-    def __init__(self, fom=None, training_set=None, validation_set=None, validation_ratio=0.1,
-                 validation_loss=None, scale_inputs=True, scale_outputs=False):
+    def __init__(self, fom=None, reduced_basis=None, training_set=None, validation_set=None, training_snapshots=None,
```
If I do not add it here it will give an error in the ```reduce``` method of the ```NeuralNetworkReductor``` when it asks whether it is None. If we do not want it here we should change that line to a ```hasattr()``` like statement.

TiPlath replied:
I am wrong. I removed it.

TiPlath commented on src/pymor/reductors/neural_network.py:
```diff
@@ -175,33 +198,45 @@ def reduce(self, hidden_layers='[(N+P)*3, (N+P)*3]', activation_function=torch.t
+            if not hasattr(self, 'validation_data'):
+                assert self.validation_snapshots is not None
+        elif self.validation_snapshots is None:
+            self.compute_validation_snapshots()
```
I tried to simplify it. I think it is more understandable now but I was not able to reduce the amount of if else logic in here. If you have a better idea, I am happy that you implement it over mine.

TiPlath commented on src/pymor/reductors/neural_network.py:
```diff
@@ -245,44 +280,63 @@ def _initialize_neural_network(self, params):
         # compute reduced basis via POD
         with self.logger.block('Building reduced basis ...'):
-            self.reduced_basis, svals = pod(U, modes=self.basis_size, rtol=self.rtol / 2.,
+            self.reduced_basis, svals = pod(self.training_snapshots, modes=self.basis_size, rtol=self.rtol / 2.,
```
I did not even think of that to be honest. Still for testing I think we want to have a fixed set of validation and data snapshots to be able to compare different architectures or other models.

TiPlath commented on src/pymor/reductors/neural_network.py:
```diff
@@ -175,33 +198,46 @@ def reduce(self, hidden_layers='[(N+P)*3, (N+P)*3]', activation_function=torch.t
+            self.validation_data = self.training_data[0:number_validation_snapshots]
+            self.validation_set = [data[0] for data in self.validation_data]
+            self.training_data = self.training_data[number_validation_snapshots:]
+
```
should we also update the ```training_set```, ```training_snapshots``` and ```validation_snapshot``` in here just to be safe for potential bugs in the future? 
Right now it only uses the ```validataion_data``` from this point onwards.

TiPlath commented on src/pymor/reductors/neural_network.py:
```diff
@@ -441,20 +518,22 @@ class NeuralNetworkStatefreeOutputReductor(NeuralNetworkReductor):
-            self.parameters_dim = fom.parameters.dim
             self.dim_output = fom.dim_output
 
         self.__auto_init(locals())
```
I am not sure about this little function. Should it stay although it is used in the super() function already?

HenKlei replied:
If you call it here again, you could remove the line with `self.validation_loss = validation_loss` and it would be safer when adding new arguments later. So I think I would keep it but remove the line above.

TiPlath commented:
@HenKlei I am done and addressed everything I think, if not please point me towards things I have not addressed. The ball is passed back to you again ;)

HenKlei commented:
> @HenKlei I am done and addressed everything I think, if not please point me towards things I have not addressed. The ball is passed back to you again ;)

Thank you very much @TiPlath! I will take a look and also make changes where I feel they are appropriate. Please focus on your thesis, I will take care of the rest here.

HenKlei commented:
@HenKlei, reminder to myself: Add @TiPlath to `AUTHORS.md`!

TiPlath commented:
> > > Could you briefly explain to me why you would like to split training and validation data already when passing them to the reductors? I do not see any benefit in doing so compared to splitting the data into a training and a validation set within the reductor.
> > 

> > 
> > As far as I can see now and remember I introduced this mainly for testing. The neural_network and neural_network_instationary test rely on checking the same train, validation parameters in different NN architectures. I also see no other benefit than the testing.
> 
> Then I would propose to remove the splitting again to make the interface a bit simpler. But maybe @sdrave also has an opinion on this?

If we remove the splitting how would you code up e.g. the demos where you want to compare different architectures and their speedup? Maybe I am wrong, but does it not matter to pass the same validation + test data to all these models to make a fair comparison?

HenKlei commented:
> > > > Could you briefly explain to me why you would like to split training and validation data already when passing them to the reductors? I do not see any benefit in doing so compared to splitting the data into a training and a validation set within the reductor.
> 
> > > As far as I can see now and remember I introduced this mainly for testing. The neural_network and neural_network_instationary test rely on checking the same train, validation parameters in different NN architectures. I also see no other benefit than the testing.
> > 
> > 
> > Then I would propose to remove the splitting again to make the interface a bit simpler. But maybe @sdrave also has an opinion on this?
> 
> If we remove the splitting how would you code up e.g. the demos where you want to compare different architectures and their speedup? Maybe I am wrong, but does it not matter to pass the same validation + test data to all these models to make a fair comparison?

In principle you are right. I am just wondering whether it is worth the additional "effort" (in terms of more complicated interfaces and so on) to ensure a fair comparison of the architectures. But I am not sure.

TiPlath commented:
> > > > > Could you briefly explain to me why you would like to split training and validation data already when passing them to the reductors? I do not see any benefit in doing so compared to splitting the data into a training and a validation set within the reductor.
> > 
> > 
> > > > As far as I can see now and remember I introduced this mainly for testing. The neural_network and neural_network_instationary test rely on checking the same train, validation parameters in different NN architectures. I also see no other benefit than the testing.
> > > 
> > > 
> > > Then I would propose to remove the splitting again to make the interface a bit simpler. But maybe @sdrave also has an opinion on this?
> > 
> > 
> > If we remove the splitting how would you code up e.g. the demos where you want to compare different architectures and their speedup? Maybe I am wrong, but does it not matter to pass the same validation + test data to all these models to make a fair comparison?
> 
> In principle you are right. I am just wondering whether it is worth the additional "effort" (in terms of more complicated interfaces and so on) to ensure a fair comparison of the architectures. But I am not sure.

true, cannot decide. lets push all responsibility to @sdrave ;)

sdrave commented:
>>    In principle you are right. I am just wondering whether it is worth the additional "effort" (in terms of more complicated interfaces and so on) to ensure a fair comparison of the architectures. But I am not sure.

> true, cannot decide. lets push all responsibility to @sdrave ;)

I think the additional effort is not that significant. So if there is some valid argument for being able to explicitly specify the validation data, I would do it.

HenKlei commented:
> > > In principle you are right. I am just wondering whether it is worth the additional "effort" (in terms of more complicated interfaces and so on) to ensure a fair comparison of the architectures. But I am not sure.
> 
> > true, cannot decide. lets push all responsibility to @sdrave ;)
> 
> I think the additional effort is not that significant. So if there is some valid argument for being able to explicitly specify the validation data, I would do it.

Thanks @sdrave for your input. Let's stick with the splitting as it is in this PR.

@TiPlath: I will take a close look into the PR in the afternoon and hope that we can merge it soon.

HenKlei commented on src/pymor/reductors/neural_network.py:
```diff
@@ -345,17 +397,31 @@ def _compute_sample(self, mu, u=None):
 
-        return [(mu, self.reduced_basis.inner(u, product=product)[:, 0])]
+        # conditional expression to check for instationary solution to return self.nt solutions
+        assert hasattr(self, 'nt')
```
```suggestion
```
I don't think this is really necessary here.

HenKlei commented on src/pymor/reductors/neural_network.py:
```diff
@@ -345,17 +397,31 @@ def _compute_sample(self, mu, u=None):
+        # conditional expression to check for instationary solution to return self.nt solutions
+        assert hasattr(self, 'nt')
+        parameters = [mu.with_(t=t) for t in np.linspace(0, self.T, self.nt)] if not self.is_stationary else [mu]
+        samples = [(mu, self.reduced_basis.inner(u_t, product=product)[:, 0]) for mu, u_t in
```
Here we actually assume that `self.reduced_basis` is an orthonormal basis of the reduced space. Since it is allowed to pass the reduced basis as an argument to the reductor now, this might not be ensured. We should either check this or rewrite this line accordingly.

To be honest, this is another line of code that feels wrong to me (please don't misunderstand me, it was the same before and I refer to the general concept how these reductors work). From my point of view, it would be nice if the reductor knew nothing about the high-dimensional states and thus also does not project them onto the reduced basis. The reduced basis would only be required for reconstruction. How do you feel about this @TiPlath? But definitely nothing to address in this PR. Maybe we should sit together (also with @sdrave and @ftschindler) and try to redesign the reductors (also having the adaptive model hierarchies in mind).

TiPlath replied:
I like the idea, happy to discuss.
To get this right, you are saying that we should split reduced basis construction and neural network approximations, such that the neural network only knows about the reduced order model? 
Thinking further, then it would also get rid of the ```fom is None``` case or am I wrong?

TiPlath replied:
Some arguments for having this inside the reductor: 
- it is called "reductor" because it should reduce the space, should we think about calling it different then?
- I like right now that I do not have to know anything about pyMOR to use these reductors because it will do all the data handling for me. I just have to pass in my simulation data. This is very (new) user friendly to me.
- ... to be continued

arguments against:
- complex class structure which does not serve a single purpose (which is not nice to have for object oriented coding)
- ... to be continued

HenKlei commented on src/pymor/reductors/neural_network.py:
```diff
@@ -423,13 +495,19 @@ class NeuralNetworkStatefreeOutputReductor(NeuralNetworkReductor):
+    training_snapshots
+        Set of solution |VectorArrays| to use for POD and training of the
+        neural network. These are the solutions to the parameters of the
+        `training_set` and can be `None` when fom is available.
```
```suggestion
        `training_set` and can be `None` when `fom` is not `None`.
```

HenKlei commented on src/pymor/reductors/neural_network.py:
```diff
@@ -423,13 +495,19 @@ class NeuralNetworkStatefreeOutputReductor(NeuralNetworkReductor):
+    validation_snapshots
+        Set of solution |VectorArrays| to use for validation in the training
+        of the neural network. These are the solutions to the parameters of
+        the `validation_set` and can be `None` when fom is available.
```
```suggestion
        the `validation_set` and can be `None` when `fom` is not `None`.
```

HenKlei commented on src/pymor/reductors/neural_network.py:
```diff
@@ -47,16 +47,23 @@ class NeuralNetworkReductor(BasicObject):
+    training_snapshots
+        |VectorArray| to use for POD and training of the
+        neural network. Contains the solutions to the parameters of the
+        `training_set` and can be `None` when FOM is available, i.e. `fom` is not `None`.
```
```suggestion
        `training_set` and can be `None` when `fom` is not `None`.
```

HenKlei commented on src/pymor/reductors/neural_network.py:
```diff
@@ -47,16 +47,23 @@ class NeuralNetworkReductor(BasicObject):
+    validation_snapshots
+        |VectorArray| to use for validation in the training
+        of the neural network. Contains the solutions to the parameters of
+        the `validation_set` and can be `None` when FOM is available, i.e. `fom` is not `None`.
```
```suggestion
        the `validation_set` and can be `None` when `fom` is not `None`.
```

HenKlei commented on src/pymor/reductors/neural_network.py:
```diff
@@ -441,20 +519,22 @@ class NeuralNetworkStatefreeOutputReductor(NeuralNetworkReductor):
-            assert len(training_set) > 0
-            self.parameters_dim = training_set[0][0].parameters.dim
-            self.dim_output = len(training_set[0][1].flatten())
+            self.dim_output = training_snapshots[0].dim
```
Is this correct? The `training_snapshots` are here still high-dimensional state snapshots and not only outputs, right?

TiPlath replied:
thanks for pointing this out, I see now that I had a big misconception here. Indeed the training_snapshots are the solutions and I initially thought solutions==outputs. Am I correct that this is not the case, i.e. outputs are the quantities of interest and not the solutions of our high dimensional states?

So would it make sense to rename validation_snapshots and training_snapshots here to training_outputs and validation_outputs and pass them into the super.__init__() as training_snapshots to work within the reduce function?

HenKlei commented on src/pymor/reductors/neural_network.py:
```diff
@@ -423,13 +495,19 @@ class NeuralNetworkStatefreeOutputReductor(NeuralNetworkReductor):
-        Set of |parameter values| to use for POD and training of the
-        neural network. If `fom` is `None`, the `training_set` has
-        to consist of pairs of |parameter values| and corresponding outputs.
+        List of |Parameter values| to use for POD and training of the
```
```suggestion
        List of |Parameter values| to use for training of the
```
That was also wrong before...

HenKlei commented on src/pymor/reductors/neural_network.py:
```diff
@@ -423,13 +495,19 @@ class NeuralNetworkStatefreeOutputReductor(NeuralNetworkReductor):
+        List of |Parameter values| to use for POD and training of the
+        neural network.
+    training_snapshots
+        Set of solution |VectorArrays| to use for POD and training of the
```
```suggestion
        Set of solution |VectorArrays| to use for training of the
```

HenKlei replied:
Is the user supposed to pass high-dimensional solutions here? That would definitely seem wrong to me...

HenKlei commented on src/pymor/reductors/neural_network.py:
```diff
@@ -466,17 +546,53 @@ def compute_training_data(self):
+            self.validation_data = []
+            for datum in self.validation_set:
+                if not self.fom:
+                    mu, output = datum
```
Now each element of the `validation_set` is already a pair containing a parameter and the output, right? That was not the case in the other reductor.

HenKlei commented on src/pymor/reductors/neural_network.py:
```diff
@@ -466,17 +546,53 @@ def compute_training_data(self):
         if output:
-            return [(mu, output.flatten())]
-        return [(mu, self.fom.output(mu).flatten())]
+            output_size = len(output)
+        else:
+            output = self.fom.output(mu).flatten()
+
+        if not self.is_stationary:
+            output_size = output.shape[0]
+            # conditional expression to check for instationary solution to return self.nt solutions
+            parameters = [mu.with_(t=t) for t in np.linspace(0, self.T, output_size)] if output_size > 1 else [mu]
+            samples = [(param, np.array([out])) for param, out in zip(parameters, output)]
+        else:
+            samples = [(mu, output)]
+        return samples
```
I don't understand this function at all. I think we have to be careful here. The output for a single mu might be vector-valued and time-dependent.

This definitely has to be adjusted.

HenKlei replied:
Looks good to me, very nice!

TiPlath commented:
@HenKlei Please check the above discussion and I think after that is resolved everything works fine.

HenKlei commented:
Hi @TiPlath,
I will take a look into this PR in the next days, but I am quite confident that we can finish this very soon.

HenKlei commented:
If I recall correctly, we discussed at some point whether to rename `training_set` to `training_parameters` (and similarly for the validation data). Do you remember what we decided on @TiPlath? To me it would make more sense to rename `set` to `parameters` since we now have `snapshots` as well and that might be confusing.

HenKlei commented on src/pymor/reductors/neural_network.py:
```diff
@@ -463,20 +555,54 @@ def compute_training_data(self):
+                self.training_data.extend(samples)
+
+    def compute_validation_data(self):
+        """Compute the training samples (the outputs to the parameters of the training set)."""
```
```suggestion
        """Compute the validation samples (the outputs to the parameters of the training set)."""
```
Here, not only the outputs are computed, but actually the pairs of parameters and outputs, right? Maybe one could slightly adjust the docstring to clarify this.

HenKlei commented on src/pymor/reductors/neural_network.py:
```diff
@@ -463,20 +555,54 @@ def compute_training_data(self):
+        """empty function to avoid computing a reduced basis."""
+
+    def compute_training_snapshots(self):
+        """empty function to avoid computing training_snapshots."""
```
```suggestion
        """Empty function since no training snapshots are required."""
```

HenKlei commented on src/pymor/reductors/neural_network.py:
```diff
@@ -463,20 +555,54 @@ def compute_training_data(self):
+                self.validation_data.extend(samples)
+
+    def compute_reduced_basis(self):
+        """empty function to avoid computing a reduced basis."""
```
```suggestion
        """Empty function since no reduced basis is required."""
```

HenKlei commented on src/pymor/reductors/neural_network.py:
```diff
@@ -463,20 +555,54 @@ def compute_training_data(self):
+        """empty function to avoid computing training_snapshots."""
+
+    def compute_validation_snapshots(self):
+        """empty function to avoid computing validation_snapshots."""
```
```suggestion
        """Empty function since no validation snapshots are required."""
```

HenKlei reviewed (changes requested):
There are some linting errors, those should be fixed to make the pipeline pass. Apart from that, one could mention the new possibilities for using the reductors also in their main docstring and in the tutorial. Moreover, you should add yourself to `AUTHORS.md` @TiPlath ;-)

TiPlath commented:
> There are some linting errors, those should be fixed to make the pipeline pass. Apart from that, one could mention the new possibilities for using the reductors also in their main docstring and in the tutorial. Moreover, you should add yourself to `AUTHORS.md` @TiPlath ;-)

what is the "main docstring"? and where do I find it? 
I could only find pymordocstring.py.

TiPlath commented on src/pymor/reductors/neural_network.py:
```diff
@@ -175,33 +199,58 @@ def reduce(self, hidden_layers='[(N+P)*3, (N+P)*3]', activation_function=torch.t
+                # concatenate the shuffled blocks into a single list
+                self.training_data[:] = [timesteps for parameter in blocks for timesteps in parameter]
+                # split training snapshots into validation and training snapshots
+                self.validation_data = self.training_data[0:number_validation_snapshots]
```
no, each block is a parameter with all its timesteps. these blocks get shuffled all together such that the blocks keep their timesteps. Then we simply concatenate the timesteps of all parameters such that the training_data has the right data format again. (e.g. for a list of blocks with 20 parameters with 11 timesteps we in the end get a list of lenght 220 where one parameter with time + a snapshot solution of this is an entry of this list.

In my understanding that should not cut a block in the middle unless we have different blocksizes (number of timesteps). does that happen?

HenKlei commented on src/pymor/reductors/neural_network.py:
```diff
@@ -245,44 +294,59 @@ def _initialize_neural_network(self, params):
+                self.validation_data = []
+                for i, (mu, u) in enumerate(validation_data_iterable):
+                    samples = self._compute_sample(mu, u.base[i * self.nt:(i + 1) * self.nt])
+                    self.validation_data.extend(samples)
```
How about this question?

HenKlei commented:
> > > There are some linting errors, those should be fixed to make the pipeline pass. Apart from that, one could mention the new possibilities for using the reductors also in their main docstring and in the tutorial. Moreover, you should add yourself to `AUTHORS.md` @TiPlath ;-)
> > 
> > 
> > what is the "main docstring"? and where do I find it? I could only find pymordocstring.py.
> 
> I mean the main docstring of the class ;-)

This hasn't happened, right? Or did I miss it?

HenKlei commented:
This is now fine from my end @TiPlath, @sdrave. Please take another brief look and then I'd suggest to finally merge this PR.

HenKlei commented on src/pymor/reductors/neural_network.py:
```diff
@@ -35,51 +35,63 @@
-    orthogonal decomposition and trains a neural network that approximates
-    the mapping from parameter space to coefficients of the full-order
-    solution in the reduced basis.
+    This is a reductor that either takes a precomputed reduced basis or
+    constructs a reduced basis using proper orthogonal decomposition
+    and trains a neural network that approximates the mapping from
```
Thanks, I was sloppy here...

HenKlei commented on src/pymor/reductors/neural_network.py:
```diff
@@ -89,20 +101,38 @@ class NeuralNetworkReductor(BasicObject):
         else:
             self.parameters_dim = fom.parameters.dim
+            if hasattr(fom, 'time_stepper'):  # instationary
+                self.nt = fom.time_stepper.nt + 1
```
I assume you miss a comment regarding the `+ 1`?

sdrave replied:
Yes.

sdrave commented on src/pymor/reductors/neural_network.py:
```diff
@@ -175,33 +205,57 @@ def reduce(self, hidden_layers='[(N+P)*3, (N+P)*3]', activation_function=torch.t
+        if self.training_data is None:
             self.compute_training_data()
+        assert self.training_data is not None
+        assert len(self.training_data) == len(self.training_parameters) * self.nt
```
The dimension of the training data should match the size of the POD basis / number of outputs.
